docs/ref-manual/cvl/invariants.md

```diff
@@ -28,29 +35,254 @@ preserved_block ::= "preserved"
+
+(invariant-assumptions)=
+Assumptions made while checking invariants
+------------------------------------------
+
+In Ethereum, the only way to change the storage state of a smart contract is
+using the smart contract's methods.  Therefore, if an invariant depends only on
+the storage of the contract, we can prove the invariant by checking it after
+calling each of the contract methods.
+
+However, it is possible to write invariants whose value depends on things other
+than the contract's storage.  The truth of an expression may depend on the
+state of other contracts or on the {term}`environment`.  For these invariants,
+the expression can change from `true` to `false` without invoking a method on
+the main contract.
+
+For example, consider the following contract:
+
+```solidity
+contract Timestamp {
+    uint256 public immutable timestamp;
+
+    constructor() {
+        timestamp = block.timestamp;
+    }
+}
 ```
 
+The following invariant will be successfully verified, although it is clearly
+false:
+
+```cvl
+invariant time_is_now(env e)
+    timestamp() == e.block.timestamp;
+```
+
+The verification is successful because the action that falsifies the invariant
+is the passage of time, rather than the invocation of a contract method.
+
+Similarly, an invariant that depends on an external contract can become false
+by calling a method on the external contract.  For example:
+
+```solidity
+contract SupplyTracker {
+    address token;
+    uint256 public supply;
+
+    constructor(address _token) {
+        token  = _token;
+        supply = token.totalSupply();
+    }
+}
+```
+
+As above, an invariant stating that `supply() == token.totalSupply()` would be
+verified, but a method on `token` might change the total supply without updating
+the `SupplyTracker` contract.  Since the Prover only checks the main contract's
+methods for preservation, it will not report that the invariant can be
+falsified.
+
+For this reason, invariants that depend on the environment or on the state of
+external contracts are a potential source of {term}`unsound`ness, and should be
+used with care.
+
+
 Filters
 -------
 
-```{todo}
-This feature is currently undocumented.
+For performance reasons, you may want to avoid checking that an invariant is
+preserved by a particular method or set of methods.  Invariant filters provide
+a method for skipping verification on a method-by-method basis.
+
+```{caution}
+Filtering out methods while checking invariants is {term}`unsound`.
 ```
 
+To filter out methods from an invariant, add a `filtered` block after the
+expression defining the invariant.  The body of the `filtered` block must
+contain a single filter of the form `var -> expr`, where `var` is a variable
+name, and `expr` is a boolean expression that may depend on `var`.
+
+If the expression evaluates to `false` with `var` replaced by a given method,
+the Prover will not check that the method preserves the invariant.  For example,
```

Comment:
the Prover will not check that the method preserves the invariant _unless there is a preserved block that specifies that method's signature (see Preserved blocks section below)._

Reply:
Why does the presence of a preserved block override the filter?  That seems confusing to me

Reply:
I think this is incorrect behavior, but have added a note to the docs